program-runtime/src/sysvar_cache.rs

```diff
@@ -23,145 +23,249 @@ impl ::solana_frozen_abi::abi_example::AbiExample for SysvarCache {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct SysvarWithAccount<T: Sysvar> {
+    value: Arc<T>,
+    account: AccountSharedData,
+}
+impl<T: Sysvar> SysvarWithAccount<T> {
+    pub fn new(value: T, account: AccountSharedData) -> Self {
+        Self {
+            value: Arc::new(value),
+            account,
+        }
+    }
+}
+impl<T: Sysvar> TryFrom<AccountSharedData> for SysvarWithAccount<T> {
+    type Error = InstructionError;
+    fn try_from(account: AccountSharedData) -> Result<Self, Self::Error> {
+        let value = bincode::deserialize(account.data())
+            .map_err(|_| InstructionError::UnsupportedSysvar)?;
```

Comment:
It gets a bit gross since this is also in `set_account`, which returns an `InstructionError::UnsupportedSysvar` for entries that aren't found... how about changing it to `InstructionError::InvalidAccountData`?

Reply:
:+1: